src/flamenco/vm/fd_vm.c

```diff
@@ -408,10 +408,22 @@ fd_vm_validate( fd_vm_t const * vm ) {
     case FD_INVALID: default: return FD_VM_ERR_INVALID_OPCODE;
     }
 
-    if( FD_UNLIKELY( instr.src_reg>10 ) ) return FD_VM_ERR_INVALID_SRC_REG; /* FIXME: MAGIC NUMBER */
-
-    int is_invalid_dst_reg = instr.dst_reg > ((validation_code == FD_CHECK_ST) ? 10 : 9); /* FIXME: MAGIC NUMBER */
-    if( FD_UNLIKELY( is_invalid_dst_reg ) ) return FD_VM_ERR_INVALID_DST_REG;
+    /* Check registers
+       https://github.com/solana-labs/rbpf/blob/v0.8.5/src/verifier.rs#L177 */
+
+    /* Source register */
+    if( FD_UNLIKELY( instr.src_reg>10 ) ) return FD_VM_ERR_INVALID_SRC_REG;
+
+    /* Special R11 register */
+    if( instr.dst_reg==11U
+        && FD_VM_SBPF_DYNAMIC_STACK_FRAMES( sbpf_version )
+        && instr.opcode.raw == 0x07
+        && ( instr.imm % FD_VM_SBPF_DYNAMIC_STACK_FRAMES_ALIGN )==0 )
+        continue;
+
+    /* Destination register. */
+    if( FD_UNLIKELY( instr.dst_reg==10U && validation_code != FD_CHECK_ST ) ) return FD_VM_ERR_INVALID_DST_REG;
+    if( FD_UNLIKELY( instr.dst_reg > 10U ) ) return FD_VM_ERR_INVALID_DST_REG;
```

Comment:
I think these destination register checks need to be conditioned on whether dynamic stack frames are enabled: https://github.com/solana-labs/rbpf/blob/c16589fcdfe8de65029a51bb28a72be0bbf29701/src/verifier.rs#L187

Reply:
no, the conditions on all registers but r11 are true for all sbpf versions.
the line you linked corresponds to the line(s) below (and FD_VM_SBPF_DYNAMIC_STACK_FRAMES is part of our check).